fastlane_bot/tools/optimizer/margpoptimizer.py

```diff
@@ -155,247 +187,288 @@ def optimize(self, sfc=None, result=None, *, params=None):
+                    print(f"[margp_optimizer] error calculating price estimates: [{e}]")
+                price_estimates_t = None
+                raise
+        
+        if P("debug"):
+            print("[margp_optimizer] price estimates = ", price_estimates_t)
+        
+        if result == self.MO_PSTART:
+            df = pd.DataFrame(price_estimates_t, index=tokens_t, columns=[targettkn])
+            df.index.name = "tknb"
+            return df
+            
+        # criterion and norm
+        crit = P("crit") or self.MOCRITR
+        assert crit in set((self.MOCRITR, self.MOCRITA)), f"crit must be {self.MOCRITR} or {self.MOCRITA}"
+        if crit == self.MOCRITA:
+            assert not pstart is None, "pstart must be provided if crit is MOCRITA"
+            assert epsaunit in pstart, f"epsaunit {epsaunit} not in pstart {P('pstart')}"
+            p_targettkn_per_epsaunit = pstart[epsaunit]/pstart[targettkn]
+            if P("debug"):
+                print(f"[margp_optimizer] 1 epsaunit [{epsaunit}] = {p_targettkn_per_epsaunit:,.4f} target [{targettkn}]")
+        crit_is_relative = crit == self.MOCRITR
+        eps_used = eps if crit_is_relative else epsa
+        eps_unit = 1 if crit_is_relative else epsaunit
+        
+        norm = P("norm") or self.MONORML2
+        assert norm in set((self.MONORML1, self.MONORML2, self.MONORMLINF)), f"norm must be MONORML1, MONORML2 or MONORMLINF [{norm}]"
+        normf = lambda x: np.linalg.norm(x, ord=norm)
+        
+        if P("verbose") or P("debug"):
+            print(f"[margp_optimizer] crit={crit} (eps={eps_used}, unit={eps_unit}, norm=L{norm})")
                 
-                # pvec is dict {tkn -> (log) price} for all tokens in p
-                pvec = {tkn: p_ for tkn, p_ in zip(tokens_t, p)}
-                pvec[targettkn] = 1
-                if P("debug") and not quiet:
-                    print(f"pvec={pvec}")
+        ## INNER FUNCTION: CALCULATE THE TARGET FUNCTION
+        def dtknfromp_f(p, *, islog10=True, asdct=False, quiet=False):
+            """
+            calculates the aggregate change in token amounts for a given price vector
+
+            :p:         price vector, where prices use the reference token as quote token
+                        this vector is an np.array, and the token order is the same as in tokens_t
+            :islog10:   if True, p is interpreted as log10(p)
+            :asdct:     if True, the result is returned as dict AND tuple, otherwise as np.array
+            :quiet:     if overrides P("debug") etc, eg for calc of Jacobian
+            :returns:   if asdct is False, a tuple of the same length as tokens_t detailing the
+                        change in token amounts for each token except for the target token (ie the
+                        quantity with target zero; if asdct is True, that same information is
+                        returned as dict, including the target token.
```

Comment:
FYI, a function returning different output types under different circumstances, requires every caller of that function to be aware of this and handle the output accordingly.
Consider splitting this function into several smaller functions, each of which returning a single (consistent) type of output, which will relieve the caller or callers from having to do that sort of special handling.